### PR TITLE
[GAIA-IR] support subgraph in ir compiler

### PR DESCRIPTION
**Pull request by shirly121**
## What do these changes do?
as titled

## Related issue number

Fixes

### CODE REVIEW

File: research/query_service/ir/compiler/src/main/java/com/alibaba/graphscope/common/jna/type/FfiVOpt.java [hunk 1]
```diff
@@ -21,7 +21,8 @@
 public enum FfiVOpt implements IntEnum<FfiVOpt> {
     Start,
     End,
-    Other;
+    Other,
+    BothV;
```
Comment (longbinlai) [inline research/query_service/ir/compiler/src/main/java/com/alibaba/graphscope/common/jna/type/FfiVOpt.java:25]: `Both`. core那边的Pr已经改成这个名字了

**Comment by shirly121** [comment]
add unit tests. https://github.com/alibaba/GraphScope/pull/1721/commits/68a281edae21a136466b90b09c2e981e6d7bacf6

**Comment by shirly121** [comment]
add unit tests https://github.com/alibaba/GraphScope/pull/1721/commits/68a281edae21a136466b90b09c2e981e6d7bacf6

File: research/query_service/ir/compiler/src/main/java/com/alibaba/graphscope/common/jna/type/FfiVOpt.java (hunk 1 above)
Comment (shirly121) [reply research/query_service/ir/compiler/src/main/java/com/alibaba/graphscope/common/jna/type/FfiVOpt.java]: done

> `Both`. core那边的Pr已经改成这个名字了

done

Comment (shirly121) [reply research/query_service/ir/compiler/src/main/java/com/alibaba/graphscope/common/jna/type/FfiVOpt.java]: > `Both`. core那边的Pr已经改成这个名字了

done

File: research/query_service/ir/compiler/src/test/java/com/alibaba/graphscope/gremlin/SubGraphStepTest.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+import java.util.Collections;
+import java.util.List;
+
+public class SubGraphStepTest {
+    private Graph graph = TinkerFactory.createModern();
+    private GraphTraversalSource g = graph.traversal();
+
+    @Test
+    public void g_E_subgraph() {
```
Comment (BingqingLyu) [inline research/query_service/ir/compiler/src/test/java/com/alibaba/graphscope/gremlin/SubGraphStepTest.java:43]: How about add a test of g_V_xxE_subgraph()?

**Review by BingqingLyu — COMMENTED** [review]
Some comments as follows:

File: research/query_service/ir/compiler/src/test/java/com/alibaba/graphscope/gremlin/antlr4/NegativeEvalTest.java [hunk 3]
```diff
@@ -113,17 +113,6 @@ public void g_V_has_key_id_test() {
         Assert.fail();
     }
 
-    @Test
-    public void g_V_bothV_test() {
```
[the remark this answers is not in the document]
Comment (shirly121) [reply research/query_service/ir/compiler/src/test/java/com/alibaba/graphscope/gremlin/antlr4/NegativeEvalTest.java:117]: it's negative test, bothV is unsupported in the past

File: research/query_service/ir/compiler/src/main/java/com/alibaba/graphscope/common/intermediate/operator/SubGraphType.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.alibaba.graphscope.common.intermediate.operator;
+
+// write subgraph to vineyard or groot
+public enum SubGraphType {
+    VINEYARD,
+    GROOT
```
[the remark this answers is not in the document]
Comment (shirly121) [reply research/query_service/ir/compiler/src/main/java/com/alibaba/graphscope/common/intermediate/operator/SubGraphType.java]: done

File: research/query_service/ir/compiler/src/test/java/com/alibaba/graphscope/gremlin/SubGraphStepTest.java (hunk 2 above)
Comment (shirly121) [reply research/query_service/ir/compiler/src/test/java/com/alibaba/graphscope/gremlin/SubGraphStepTest.java:43]: done

File: research/query_service/ir/compiler/src/test/java/com/alibaba/graphscope/gremlin/antlr4/PositiveEvalTest.java [hunk 5]
```diff
@@ -871,4 +871,14 @@ public void g_V_out_endV_test() {
...
     }
+
+    @Test
+    public void g_E_bothV_test() {
+        Assert.assertEquals(g.E().bothV(), eval("g.E().bothV()"));
+    }
+
+    @Test
+    public void g_E_subgraph_test() {
+        Assert.assertEquals(g.V().subgraph("graph_X"), eval("g.V().subgraph(\"graph_X\")"));
```
[the remark this answers is not in the document]
Comment (shirly121) [reply research/query_service/ir/compiler/src/test/java/com/alibaba/graphscope/gremlin/antlr4/PositiveEvalTest.java]: done

**Pull request merged**